### PR TITLE
[Flight] Encode React Elements in Replies as Temporary References

### PR DESCRIPTION
**Pull request by sebmarkbage**
Currently you can accidentally pass React Element to a Server Action. It warns but in prod it actually works because we can encode the symbol and otherwise it's mostly a plain object. It only works if you only pass host components and no function props etc. which makes it potentially error later. The first thing this does it just early hard error for elements.

I made Lazy work by unwrapping though since that will be replaced by Promises later which works.

Our protocol is not fully symmetric in that elements flow from Server -> Client. Only the Server can resolve Components and only the client should really be able to receive host components. It's not intended that a Server can actually do something with them other than passing them to the client.

In the case of a Reply, we expect the client to be stateful. It's waiting for a response. So anything we can't serialize we can still pass by reference to an in memory object. So I introduce the concept of a TemporaryReferenceSet which is an opaque object that you create before encoding the reply. This then stashes any unserializable values in this set and encode the slot by id. When a new response from the Action then returns we pass the same temporary set into the parser which can then restore the objects. This lets you pass a value by reference to the server and back into another slot.

For example it can be used to render children inside a parent tree from a server action:

```
export async function Component({ children }) {
  "use server";
  return <div>{children}</div>;
}
```

(You wouldn't normally do this due to the waterfalls but for advanced cases.)

A common scenario where this comes up accidentally today is in `useActionState`.

```
export function action(state, formData) {
  "use server";
   if (errored) {
     return <div>This action <strong>errored</strong></div>;
   }
   return null;
}
```

```
const [errors, formAction] = useActionState(action);
return <div>{errors}<div>;
```

It feels like I'm just passing the JSX from server to client. However, because `useActionState` also sends the previous state *back* to the server this should not actually be valid. Before this PR this actually worked accidentally. You get a DEV warning but it used to work in prod. Once you do something like pass a client reference it won't work tho. We could perhaps make client references work by stashing where we got them from but it wouldn't work with all possible JSX.

By adding temporary references to the action implementation this will work again - on the client. It'll also be more efficient since we don't send back the JSX content that you shouldn't introspect on the server anyway.

However, a flaw here is that the progressive enhancement of this case won't work because we can't use temporary references for progressive enhancement since there's no in memory stash. What is worse is that it won't error if you hydrate. ~It also will error late in the example above because the first state is "undefined" so invoking the form once works - it errors on the second attempt when it tries to send the error state back again.~ It actually errors on the first invocation because we need to eagerly serialize "previous state" into the form. So at least that's better.

I think maybe the solution to this particular pattern would be to allow JSX to serialize if you have no temporary reference set, and remember client references so that client references can be returned back to the server as client references. That way anything you could send from the server could also be returned to the server. But it would only deopt to serializing it for progressive enhancement. The consequence of that would be that there's a lot of JSX that might accidentally seem like it should work but it's only if you've gotten it from the server before that it works. This would have to have pair them somehow though since you can't take a client reference from one implementation of Flight and use it with another.

### CODE REVIEW

File: packages/react-client/src/ReactFlightReplyClient.js [hunk 1]
```diff
@@ -288,23 +380,19 @@ export function processReply(
...
-            'Classes or null prototypes are not supported.',
+        if (temporaryReferences === undefined) {
+          throw new Error(
+            'Only plain objects, and a few built-ins, can be passed to Server Actions. ' +
+              'Classes or null prototypes are not supported.',
+          );
+        }
+        // We can serialize class instances as temporary references.
+        return serializeTemporaryReferenceID(
+          writeTemporaryReference(temporaryReferences, value),
```
Comment (sebmarkbage) [inline packages/react-client/src/ReactFlightReplyClient.js:391]: Anything we can't serialize, we can still encode as a temporary reference.

Unfortunately, for some things like the stuff below, we only warn in DEV that we can't serialize it so for any of those, we don't put it as a temporary reference.

So any of those are really invalid in both cases.

**Comment by unstubbable** [comment]
Lovely PR description! ❤️

File: packages/react-client/src/ReactFlightClient.js [hunk 2]
```diff
@@ -689,6 +694,18 @@ function parseModelString(
         const metadata = getOutlinedModel(response, id);
         return createServerReferenceProxy(response, metadata);
       }
+      case 'T': {
+        // Temporary Reference
+        const id = parseInt(value.slice(2), 16);
+        const temporaryReferences = response._tempRefs;
+        if (temporaryReferences == null) {
```
Comment (shuding) [inline packages/react-client/src/ReactFlightClient.js:701]: Saw that the type is `void | TemporaryReferenceSet` and in the other place we're checking with `=== undefined`. Is the `== null` expected?

File: packages/react-client/src/ReactFlightReplyClient.js [hunk 3]
```diff
@@ -219,14 +307,18 @@ export function processReply(
...
           partValue => {
-            const partJSON = JSON.stringify(partValue, resolveToJSON);
-            // $FlowFixMe[incompatible-type] We know it's not null because we assigned it above.
-            const data: FormData = formData;
-            // eslint-disable-next-line react-internal/safe-string-coercion
-            data.append(formFieldPrefix + promiseId, partJSON);
-            pendingParts--;
-            if (pendingParts === 0) {
-              resolve(data);
+            try {
```
Comment (gnoff) [inline packages/react-client/src/ReactFlightReplyClient.js:310]: was this supposed to always be wrapped in try? I don't quite follow why this change required the try now but it didn't before

File: packages/react-client/src/ReactFlightReplyClient.js [hunk 4]
```diff
@@ -206,6 +216,84 @@ export function processReply(
...
+          const init = lazy._init;
+          if (formData === null) {
+            // Upgrade to use FormData to allow us to stream this value.
+            formData = new FormData();
+          }
+          try {
+            const resolvedModel = init(payload);
+            // We always outline this as a separate part even though we could inline it
+            // because it ensures a more deterministic encoding.
+            pendingParts++;
```
Comment (gnoff) [inline packages/react-client/src/ReactFlightReplyClient.js:245]: why isn't this lifted above the try and then not re-incremented in the catch? I assume nothing can throw between here and the decrement but if it did we'd end up with a double increment in the catch if it was a thenable.

mostly I'm asking b/c I think you did this on purpose and I want to know why, not that I think it's wrong per se

**Review by gnoff — APPROVED** [review]

File: packages/react-client/src/ReactFlightClient.js (hunk 2 above)
Comment (sebmarkbage) [reply packages/react-client/src/ReactFlightClient.js:701]: Yea, I just do it since it's a public API and a user might still pass `null` incorrectly but really we should probably be more strongly opinionated about not doing that.

File: packages/react-client/src/ReactFlightReplyClient.js (hunk 3 above)
Comment (sebmarkbage) [reply packages/react-client/src/ReactFlightReplyClient.js:310]: Yes, I think it was an oversight because the .then reject wouldn't catch the case when a promise's body wasn't able to be serialized and threw.

File: packages/react-client/src/ReactFlightReplyClient.js (hunk 4 above)
Comment (sebmarkbage) [reply packages/react-client/src/ReactFlightReplyClient.js:245]: We need to keep it incremented when serializing the inner tree since otherwise the inner one can complete if it finishes. There are branches below that doesn't increment so it would require them to remember to decrement.

However I think it's a big bug because resolveToJSON *can* throw in this mode (unlike FlightServer). Which is a bit strange by itself because it means that we reject creating a payload in this case where as other Flight throws on the receiving side.

This means that if we later shifted this to be streaming (we don't right now mainly because Safari doesn't support it in fetch but also for simplicity), we'd need to change the error handling semantics.

File: packages/react-client/src/ReactFlightReplyClient.js [hunk 5]
```diff
@@ -206,6 +216,84 @@ export function processReply(
...
+              typeof x === 'object' &&
+              x !== null &&
+              typeof x.then === 'function'
+            ) {
+              // Suspended
+              pendingParts++;
+              const lazyId = nextPartId++;
+              const thenable: Thenable<any> = (x: any);
+              thenable.then(
+                partValue => {
```
Comment (sebmarkbage) [inline packages/react-client/src/ReactFlightReplyClient.js:265]: Writing a test I found another bug. The resolved value of the thrown promise is not the same as the value if you try it again.
